ComboboxControl and FormTokenField: enhance components with custom render callback for options

manzoorwanijk opened the pull request:
## What?
This PR
- Adds an optional render callback to the `SuggestionsList` and `ComboboxControl` components in order to allow the consumer to control how an option is rendered inside `SuggestionsList` for a combobox or any other consumer of `SuggestionsList`
- Adds a corresponding story to `ComboboxControl` to demonstrate the new behavior

## Why?
In many use cases of combobox, we need to show more detail to the user for an option in the list apart from just the label. More details help users identify the correct option from the list when there are plenty of them available with similar labels. 

## How?
- It first adds an optional (thus full backward compatible) prop `renderSuggestion` to `SuggestionsList` component.
- If `renderSuggestion` is passed, it will be given preference and the suggestion will be passed to it to allow it to render the suggestion.
- It also adds `renderOption` prop to `ComboboxControl` which in turn gets passed to `SuggestionsList` above.

## Testing Instructions

- Checkout the branch
- Run `npm run storybook:dev`
- Check combobox story in Storybook
- Confirm that the existing story behaves as before i.e. it is backward compatible
- Confirm that `With Render Option` story shows more details in the options for combobox like Age, Country.

## Screenshots or screencast

<img width="1044" alt="Screenshot 2022-07-21 at 5 49 24 PM" src="https://user-images.githubusercontent.com/18226415/180211982-709e98a0-779e-4adf-9924-7b629c664d3b.png">



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202011579628273

manzoorwanijk committed:
Add renderSuggestion prop to SuggestionsList component

manzoorwanijk committed:
Add renderOption prop to ComboboxControl component

manzoorwanijk committed:
Add story to ComboboxControl component

ciampo commented on packages/components/src/form-token-field/suggestions-list.tsx:
```diff
@@ -122,6 +123,24 @@ export function SuggestionsList< T extends string | { value: string } >( {
+				let output: ReactNode;
+
+				if ( typeof renderSuggestion === 'function' ) {
+					output = renderSuggestion( suggestion );
```
If we pass `suggestion` as a property, this would enable us to pass more data in the future without introducing breaking changes to this function's signature (for example, we may want to pass `matchText` or any other piece of data)

```suggestion
					output = renderSuggestion( { suggestion } );
```

WDYT?

manzoorwanijk replied:
Yes, I always prefer and like that object like approach but I saw that `displayTransform` callback has a similar signature, thus I made it the same.

ciampo replied:
I see, that makes sense. I'm still tempted to go for the suggested object-like approach (and potentially look into introducing that same signature to the `displayTransform` at some point) — @mirka , what are your thoughts here?

manzoorwanijk replied:
Yes, I have changed it to object-like approach.

ciampo reviewed (commented):
Hey @manzoorwanijk , thank you for the contribution!

The code changes are clear, and I definitely understand the need for this change.

The only reservation that I currently have is around API design. We're currently in the process of [investigating](https://github.com/WordPress/gutenberg/pull/41726) how to best refactor `CustomSelectControl` using `ariakit`, and we'll very likely apply the same treatment to `ComboboxControl` and related. Therefore, I'm going to take some additional time to understand if introducing the `renderSuggestion` prop is compatible with those potential changes.

Apart from that, here's a few additional notes from this first round of review:

- Should we add the same prop to the `FormTokenField` components?
- We should add documentation about this prop to the `ComboboxControl` README (and potentially `FormTokenField`, in case)
- We should add a [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) entry

manzoorwanijk committed:
Improve renderSuggestion and renderOption signature

manzoorwanijk committed:
Add renderSuggestion prop to FormTokenField

manzoorwanijk committed:
Add props to README

manzoorwanijk committed:
Update changelog

manzoorwanijk commented:
> * Should we add the same prop to the `FormTokenField` components?
> * We should add documentation about this prop to the `ComboboxControl` README (and potentially `FormTokenField`, in case)
> * We should add a [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) entry

@ciampo Thank you for the feedback. I agree with your suggestions. I have made the changes accordingly.

manzoorwanijk committed:
Fix fatal error

rcanepa commented on packages/components/src/combobox-control/index.js:
```diff
@@ -279,6 +280,9 @@ function ComboboxControl( {
 							onHover={ setSelectedSuggestion }
 							onSelect={ onSuggestionSelected }
 							scrollIntoView
+							renderSuggestion={ ( { suggestion } ) =>
+								renderOption?.( { option: suggestion } )
+							}
```
I believe this is breaking the default case where no `renderOption` prop is passed. In concrete, `renderSuggestion` is always a function regardless of the value of `renderOption`, which means that `if ( typeof renderSuggestion === 'function' ) {` (located at line 128 in packages/components/src/form-token-field/suggestions-list.tsx) will be always true.

In the end, the default case displays an empty string for each option:
![image](https://user-images.githubusercontent.com/3418513/181057009-15847f36-9829-4197-a22b-730e8e692479.png)

manzoorwanijk replied:
Oh yeah, nice catch! Thanks @rcanepa

manzoorwanijk replied:
Pushed a fix.

manzoorwanijk committed:
Fix backward compatibility

rcanepa commented on packages/components/src/combobox-control/index.js:
```diff
@@ -280,8 +280,14 @@ function ComboboxControl( {
+							renderSuggestion={
+								typeof renderOption === 'function'
+									? ( { suggestion } ) => {
+											renderOption( {
```
This is almost ready. Now, we need to add a `return` here. Otherwise, the anonymous function doesn't return anything, and the `With Render Option` case does not work.

manzoorwanijk replied:
🤦‍♂️ Thanks

rcanepa commented on packages/components/src/combobox-control/index.js:
```diff
@@ -279,6 +280,15 @@ function ComboboxControl( {
+							renderSuggestion={
+								typeof renderOption === 'function'
+									? ( { suggestion } ) => {
+											renderOption( {
```
```suggestion
											return renderOption( {
```

manzoorwanijk committed:
Update packages/components/src/combobox-control/index.js

Co-authored-by: Renzo Canepa <rcanepag@gmail.com>

rcanepa reviewed (approved):
LGTM now.

I don't see any regression anymore, and the new `renderOption` prop works as expected.

ciampo commented:
I would request that we wait before merging, as explained in an [earlier comment](https://github.com/WordPress/gutenberg/pull/42597#pullrequestreview-1048149025).

Since the changes in this PR would be making changes to the public APIs of a few components, we'd like to make sure that these changes are compatible with the future changes that we are planning for these exact components, as introducing breaking changes is something that we try to avoid as much as possible for the `@wordpress/components` package.

Does the proposed changes have a high priority ? I'm asking because I'm quite busy these days and I need to juggle priorities in my queues.

manzoorwanijk commented:
> I would request that we wait before merging, as explained in an [earlier comment](https://github.com/WordPress/gutenberg/pull/42597#pullrequestreview-1048149025).

I am not sure what it's supposed to look like in the future.

> Does the proposed changes have a high priority?

Yes, we have a need for this change in order to use the component inside wp-admin for Jetpack.

ciampo commented:
> I am not sure what it's supposed to look like in the future.

That's ok, it's the job of the components squad (mainly @mirka and myself) to understand that.

> Yes, we have a need for this change in order to use the component inside wp-admin for Jetpack.

I'll try to prioritize this task, then. Thank you!

ciampo commented on packages/components/src/combobox-control/stories/index.js:
```diff
@@ -287,3 +287,72 @@ _default.args = {
+	...details,
+} ) );
+
+export const WithRenderOption = ( args ) => {
```
I noticed that, in this Storybook example, even when there's no value set, the "X" button is shown in the right part of the input

![Screenshot 2022-08-02 at 13 34 06](https://user-images.githubusercontent.com/1083581/182365184-4be4a6c3-ccfa-45e7-8281-1536d61cdfb1.png)

manzoorwanijk replied:
Because `allowReset` prop in `ComboboxControl` is set to `true` by default.

ciampo reviewed (commented):
Sorry for the delay in answering, and thank you for your patience!

**TL;DR:** let's rename `renderProp` to `__experimentalRenderItem` on both components. I believe @mirka also had some further feedback to leave, so let's wait for that too.

---

More details:

After discussing the matter with @mirka , we analysed the main options that we have in terms of API surface for `CustomSelectControl` (and, similarly, `ComboboxControl` and `ForkTokenField`). While it's still early for us to commit "officially", we decided to explore the last approach highlighted in [this comment](https://github.com/WordPress/gutenberg/pull/41952#issuecomment-1171140086):

- it is the most similar to the current approach, and therefore the least disruptive
- since we don't want to fully commit yet, let's add an `__experimental` prefix in front (as we continue the work on `CustomSelectControl`, we will either stabilize these props, or remove them in favor of a different approach)
- in order to have the highest flexibility, we may want to use two separate props (depending on the component):
  - one for rendering each option in the dropdown (called `__experimentalRenderItem`)
  - another one for rendering the selected value(s) that are shown also when the component is "closed"  (called `__experimentalRenderValue`)
- as we stabilize these props, we may need to make some changes / expand what the `options` prop is capable of (e.g accept option groups, how to pass custom additional data), which may result into potential **breaking changes**  (and therefore I would recommend that you follow the work on these components closely to avoid any disruption)

---

Question: in the current state of this PR, would you expect to be able to find the items by searching for their country / age as well ? The potential need for matching the search string against _all data fields_ (and not just the value)

manzoorwanijk committed:
Add __experimental prefix to render callbacks

manzoorwanijk commented:
> **TL;DR:** let's rename `renderProp` to `__experimentalRenderItem` on both components. I believe @mirka also had some further feedback to leave, so let's wait for that too.

Thank you for the feedback @ciampo . I have made the changes and used `__experimentalRenderItem` for all the components involved to make the usage consistent.

> in order to have the highest flexibility, we may want to use two separate props (depending on the component):
> * one for rendering each option in the dropdown (called `__experimentalRenderItem`)
> * another one for rendering the selected value(s) that are shown also when the component is "closed"  (called `__experimentalRenderValue`)

I am not sure if `__experimentalRenderValue` will work as expected because it will affect the current value of the input being entered by user while searching for the options.

> Question: in the current state of this PR, would you expect to be able to find the items by searching for their country / age as well ? The potential need for matching the search string against _all data fields_ (and not just the value)

No, that's not currently possible. Maybe in some future iteration, we can do that.

mirka commented on packages/components/src/form-token-field/types.ts:
```diff
@@ -154,6 +158,10 @@ export interface FormTokenFieldProps
+	__experimentalRenderItem?: ( args: { item: string } ) => ReactNode;
 }
 
 export interface SuggestionsListProps< T = string | { value: string } > {
```
I believe this now needs to be an inexact object:

```suggestion
export interface SuggestionsListProps<
	T = string | ( Record< string, unknown > & { value: string } )
> {
```

manzoorwanijk replied:
I agree. Thanks

mirka commented on packages/components/src/combobox-control/stories/index.js:
```diff
@@ -287,3 +287,76 @@ _default.args = {
 	allowReset: false,
 };
+
+const authors = [
```
This story file wasn't in great shape to begin with, so I'd like to refactor and simplify it a bit, if that's ok.

<details>
<summary>Click to expand diff</summary>

```diff
diff --git a/packages/components/src/combobox-control/stories/index.js b/packages/components/src/combobox-control/stories/index.js
index b713456e98..3c6974e954 100644
--- a/packages/components/src/combobox-control/stories/index.js
+++ b/packages/components/src/combobox-control/stories/index.js
@@ -266,7 +266,7 @@ const mapCountryOption = ( country ) => ( {
 
 const countryOptions = countries.map( mapCountryOption );
 
-function CountryCodeComboboxControl( args ) {
+function Template( args ) {
 	const [ value, setValue ] = useState( null );
 
 	return (
@@ -275,88 +275,59 @@ function CountryCodeComboboxControl( args ) {
 				{ ...args }
 				value={ value }
 				onChange={ setValue }
-				label="Select a country"
-				options={ countryOptions }
 			/>
 			<p>Value: { value }</p>
 		</>
 	);
 }
-export const _default = CountryCodeComboboxControl.bind( {} );
-_default.args = {
+export const Default = Template.bind( {} );
+Default.args = {
 	__next36pxDefaultSize: false,
 	allowReset: false,
+	label: 'Select a country',
+	options: countryOptions,
 };
 
-const authors = [
+const authorOptions = [
 	{
-		name: 'Hermann P. Schnitzel',
-		age: 45,
-		country: 'Germany',
-	},
-	{
-		name: 'Shequondolisa Bivouac',
-		age: 43,
-		country: 'France',
-	},
-	{
-		name: 'Bodrum Salvador',
-		age: 42,
-		country: 'Spain',
-	},
-	{
-		name: 'Parsley Montana',
+		value: 'parsley',
+		label: 'Parsley Montana',
 		age: 48,
 		country: 'Germany',
 	},
 	{
-		name: 'Cabbage New York',
+		value: 'cabbage',
+		label: 'Cabbage New York',
 		age: 44,
 		country: 'France',
 	},
 	{
-		name: 'Jake Weary',
+		value: 'jake',
+		label: 'Jake Weary',
 		age: 41,
 		country: 'United Kingdom',
 	},
 ];
 
-const authorOptions = authors.map( ( { name, ...details } ) => ( {
-	value: name,
-	label: name,
-	...details,
-} ) );
-
-export const WithRenderOption = ( args ) => {
-	const [ value, setValue ] = useState( '' );
-
-	return (
-		<>
-			<ComboboxControl
-				{ ...args }
-				value={ value }
-				onChange={ setValue }
-				label="Select an author"
-				options={ authorOptions }
-				__experimentalRenderItem={ ( { item } ) => {
-					const { label, age, country } = item;
-					return (
-						<div>
-							<div style={ { marginBottom: '0.2rem' } }>
-								{ label }
-							</div>
-							<small>
-								Age: { age }, Country: { country }
-							</small>
-						</div>
-					);
-				} }
-			/>
-			<p>Selected author: { value }</p>
-		</>
-	);
-};
-
-WithRenderOption.args = {
-	allowReset: false,
+/**
+ * The rendered output of each suggestion can be customized by passing a
+ * render function to the `__experimentalRenderItem` prop. (This is still an experimental feature
+ * and is subject to change.)
+ */
+export const WithCustomRenderItem = Template.bind( {} );
+WithCustomRenderItem.args = {
+	...Default.args,
+	label: 'Select an author',
+	options: authorOptions,
+	__experimentalRenderItem: ( { item } ) => {
+		const { label, age, country } = item;
+		return (
+			<div>
+				<div style={ { marginBottom: '0.2rem' } }>{ label }</div>
+				<small>
+					Age: { age }, Country: { country }
+				</small>
+			</div>
+		);
+	},
 };
```

</details>

manzoorwanijk replied:
Updated. Thank you.

mirka commented on packages/components/src/form-token-field/README.md:
```diff
@@ -56,6 +56,7 @@ The `value` property is handled in a manner similar to controlled form component
     -   `removed` - The user removed an existing token.
     -   `remove` - The user focused the button to remove the token.
     -   `__experimentalInvalid` - The user tried to add a token that didn't pass the validation.
+-   `renderSuggestion` - Custom renderer for suggestions.
```
```suggestion
-   `__experimentalRenderItem` - Custom renderer for suggestions.
```

manzoorwanijk replied:
Done, thanks

mirka commented on packages/components/src/combobox-control/stories/index.js:
```diff
@@ -287,3 +287,76 @@ _default.args = {
+
+WithRenderOption.args = {
+	allowReset: false,
+};
```
Let's also add this quick story to `form-token-field/stories/index.tsx`:

```tsx
/**
 * The rendered output of each suggestion can be customized by passing a
 * render function to the `__experimentalRenderItem` prop. (This is still an experimental feature
 * and is subject to change.)
 */
export const WithCustomRenderItem: ComponentStory< typeof FormTokenField > =
	DefaultTemplate.bind( {} );
WithCustomRenderItem.args = {
	...Default.args,
	__experimentalRenderItem: ( { item } ) => (
		<div>{ `${ item } — a nice place to visit` }</div>
	),
};
```

manzoorwanijk replied:
Added, thank you.
<img width="805" alt="Screenshot 2022-08-03 at 5 51 12 PM" src="https://user-images.githubusercontent.com/18226415/182606380-6c492e2c-f2d6-4272-9110-7e220cf568fb.png">

manzoorwanijk committed:
Update packages/components/src/form-token-field/types.ts

Co-authored-by: Lena Morita <lena@jaguchi.com>

manzoorwanijk committed:
Improve and clean up story

manzoorwanijk committed:
Update docs

manzoorwanijk committed:
Update form-token-field story

manzoorwanijk committed:
Merge branch 'WordPress:trunk' into add/render-cb-to-suggestions-list

mirka reviewed (commented):
Thanks for the quick turnaround! The changes are good on my end. Once Marco gives the final 👍 we're good to go.

ciampo commented on packages/components/src/combobox-control/README.md:
```diff
@@ -116,6 +116,13 @@ The current value of the input.
 -   Type: `mixed`
 -   Required: Yes
 
+#### \_\_experimentalRenderItem
```
Can we omit the `\` characters to escape the underscores? They would prevent this line of text from coming up when searching for the `__experimentalRenderItem` prop

ciampo commented on packages/components/src/combobox-control/README.md:
```diff
@@ -116,6 +116,13 @@ The current value of the input.
 
+#### \_\_experimentalRenderItem
+
+Custom renderer for options in suggestion list.
```
Following up from the previous comment:

```suggestion
Custom renderer invoked for each option in the suggestion list. The render prop receives as its argument an object containing, under the `item` key, the single option's data (directly from the array of data passed to the `options` prop).
```

ciampo commented on packages/components/src/combobox-control/README.md:
```diff
@@ -116,6 +116,13 @@ The current value of the input.
+
+Custom renderer for options in suggestion list.
+
+-   Type: `Function` - `( args: { option: object } ) => ReactNode`
```
```suggestion
-   Type: `Function` - `( args: { item: object } ) => ReactNode`
```

ciampo commented on packages/components/src/form-token-field/README.md:
```diff
@@ -56,6 +56,7 @@ The `value` property is handled in a manner similar to controlled form component
     -   `removed` - The user removed an existing token.
     -   `remove` - The user focused the button to remove the token.
     -   `__experimentalInvalid` - The user tried to add a token that didn't pass the validation.
+-   `__experimentalRenderItem` - Custom renderer for suggestions.
```
Let's add some more details here (similarly to `ComboboxControl`'s README)

ciampo commented on packages/components/src/form-token-field/types.ts:
```diff
@@ -154,9 +158,15 @@ export interface FormTokenFieldProps
 
-export interface SuggestionsListProps< T = string | { value: string } > {
+export interface SuggestionsListProps<
+	T = string | ( Record< string, unknown > & { value: string } )
```
Maybe we could a comment here about the fact that `T` can be either a `string` or `Record< string, unknown > & { value: string }` because of its usage in both `FormTokenField` (as a `string`) and `ComboboxControl` (as an object with the `value` prop) (or at least, this is what I'm gathering from the code)

ciampo commented on packages/components/src/combobox-control/stories/index.js:
```diff
@@ -275,15 +275,59 @@ function CountryCodeComboboxControl( args ) {
+	},
+	{
+		value: 'jake',
+		label: 'Jake Weary',
```
[lol](https://jquery.com/)

ciampo reviewed (approved):
LGTM 🚀  Feel free to merge once the few remaining pieces of feedback are addressed!

> I am not sure if `__experimentalRenderValue` will work as expected because it will affect the current value of the input being entered by user while searching for the options.

I though what we could still use the results of `__experimentalRenderValue` while the search input is not focused, but that would required completely replacing the `input` element with a custom component showing the selected value(s).

Let's leave it for now, and re-evaluate once we work on the overhaul of these components. Another thing that I'd like to do is to "normalize" the shape of `options` (for `ComboboxControl`) and `suggestions` (for `FormTokenField`) to have the same "type", and to make it clean that, whatever that type it, it's going to be the same type that `__experimentalRenderItem` and `__experimentalRenderValue` expect

> > Question: in the current state of this PR, would you expect to be able to find the items by searching for their country / age as well ? The potential need for matching the search string against _all data fields_ (and not just the value)
> 
> No, that's not currently possible. Maybe in some future iteration, we can do that.

Apologies in case my question got misunderstood — I was asking mostly to understand your expectations as a user of the component. I'm not expecting that this PR introduces that functionality — we're still at the phase where we're evaluating what data should be "searchable" to the user. It's not trivial since, with the introduction of the render props, the component doesn't really know what data is shown to the user (and therefore what the user may deem "searchable")

manzoorwanijk committed:
Improve docs

manzoorwanijk commented:
> I was asking mostly to understand your expectations as a user of the component.

Yes, in the future, it would be great to have the values being searchable by anything that is rendered in the suggestions list.

manzoorwanijk commented:
> Feel free to merge once the few remaining pieces of feedback are addressed!

Thank you for the feedback @ciampo. I have addressed all the feedback. Feel free to merge it when ready. I am not authorized to merge PRs here.

manzoorwanijk committed:
Merge branch 'WordPress:trunk' into add/render-cb-to-suggestions-list

manzoorwanijk committed:
Merge branch 'WordPress:trunk' into add/render-cb-to-suggestions-list
